File: matlab/+smalltargetmotiondetectors/test/test_model/test_FracSTMD.m
% Test script for FracSTMD model

% Clear command window, workspace, and close all figures
clc, clear, close all;

% Add necessary paths
filePath = mfilename('fullpath');
indexPath = strfind(filePath, 'Small-Target-Motion-Detectors');
addpath(filePath(1:indexPath(end)+35));

% Import necessary packages
import smalltargetmotiondetectors.*;
import smalltargetmotiondetectors.api.*;
import smalltargetmotiondetectors.tool.*;

%% Model instantiation

% Instantiate the FracSTMD model
model = instancing_model('FracSTMD');

%% Input source

% Define input source (e.g., image stream or video stream)

% Example 1: Image stream
% hSteam = ImgstreamReader();

% Example 2: Image stream from demo data
% hSteam = ImgstreamReader( ...
%     [filePath(1:indexPath(end)+28),'/demodata/DemoFig*.tif'], ...
%      10, 100 );

% Example 3: Image stream from real-world scene
% hSteam = ImgstreamReader( ...
%     ['I:/Dataset/STMD_Dataset/Real-World-Scence-Material/RIST/', ...
%     'GX010241-1/Real-Image*.jpg'], ...
%     100, 1000 );

% Example 4: Video stream from real-world scene
hSteam = VidstreamReader( ...
    ['D:/Dataset/STMD_Dataset/Real-World-Scence-Material/RIST/', ...
     'GX010290-1/GX010290-1.mp4']);

% Example 5: Image stream from simulated data
% hSteam = ImgstreamReader( ...
%     ['I:/Dataset/STMD_Dataset/Simulated-DataSet/White-Background/', ...
%     'BV-250-Leftward/SingleTarget-TW-5-TH-5-TV-250-TL-0-Rightward', ...
%     '-Amp-0-Theta-0-TemFre-2-SamFre-200/', ...
%     'GeneratingDataSet*.tif'], ...
%     100, 1000 );

% Example 6: Image stream with a single image
% hSteam = ImgstreamReader( ...
%     ['D:/Dataset/STMD_Dataset/Real-World-Scence-Material/RIST/', ...
%     'GX010290-1/Real-Image*.jpg'], ...
%     1, 4000 );

% Instantiate visualization handle
hVisual = get_visualize_handle(class(model));
% Visualization(class(model));

% Initialize the model
model.init();

%% Run inference
while hSteam.hasFrame && hVisual.hasFigHandle
    % Get the next frame from the input source
    [grayImg, colorImg] = hSteam.get_next_frame();
    
    % Perform inference using the model
    result = inference(model, grayImg);
    
    % Visualize the result
    hVisual.show_result(colorImg, result);
end

File: python/smalltargetmotiondetectors/test/test_model/test_FracSTMD.py
import os
import sys

# Get the full path of this file
filePath = os.path.realpath(__file__)
# Find the index of '/+smalltargetmotiondetectors/'
indexPath = filePath.rfind('smalltargetmotiondetectors')
# Add the path to the package containing the models
import_path = filePath[:indexPath]
sys.path.append(import_path)

from smalltargetmotiondetectors.api import *
from smalltargetmotiondetectors.util.iostream import *

''' Model instantiation '''
objModel = instancing_model('FracSTMD')

''' Input '''
# Demo video (RIST)
hSteam = VidstreamReader(os.path.join(filePath[:indexPath-7], 'demodata', 'simulatedVideo0_orignal_1000Hz.mp4'))
# hSteam = VidstreamReader(os.path.join(filePath[:indexPath-7], 'demodata', 'RIST_GX010290_orignal_240Hz.mp4'))

''' Get visualization handle '''
hVisual = get_visualize_handle(objModel.__class__.__name__)

''' Initialize the model '''
# set the parameter list
objModel.set_parameter()
# print the parameter list
objModel.print_parameter()
# init
objModel.init_config()

# Run inference
while hSteam.hasFrame and hVisual.hasFigHandle:
    # Get the next frame from the input source
    grayImg, colorImg = hSteam.get_next_frame()
    
    # Perform inference using the model
    result = inference(objModel, grayImg)
    
    # Visualize the result
    hVisual.show_result(colorImg, result)
